### PR TITLE
Delegation to workers

### PR DESCRIPTION
**Pull request by gares**
We delegate with fork or via marshaling on non-Unix.

TODO:
- [x] interruption of workers
  - [ ] test
- [x] invalidation of the job's queue
- [x] catch Unix_error/IO_error everywhere
- [ ] test on windows (for now non-fork based code only tested on Unix)
- [ ] decide or give user options about states being sent back (currently fat, tested with no state as well)
- [ ] feedback about ongoing work?
- [ ] the end of `ExecutionManager.observe` is "ready" to even have master running asynchronusly. We could even try `Lwt.detach` with a mutex around observe (the former observe would be non-blocking, the second one blocking.. could be refined with a queue). To be understood if things like LSP queries could be performed in "parallel" with observe (e.g. hovering). This could also pave the way to "interruption" by actually editing and not via a signal, but fairness may not be there...
- [ ] I did not write code that looks at a proof and decides if it is OK and worth delegating it. I don't recall where these two decisions belong. The former could be in scheduler.

I'm also wondering if we should have some configurations on the user side, at least in an early stage, to see what works the best:
- check proofs automatically
  - no (on demand by hovering on it + click)
  - yes
    - immediately (old good PG)
    - delegate (like CoqIDE or VSCoq pre LSP)
       - how many workers
 - keep proof states to speed up editing/repair of proofs (consumes more ram, can be applied to all models above)

### CODE REVIEW

File: ide/vscoq/vscoqtop.ml [hunk 1]
```diff
@@ -11,285 +11,33 @@
...
-    let hyp = `Assoc ([
-      "identifiers", `List ids;
-      "type", `String (Pp.string_of_ppcmds typ);
-      "diff", `String "None";
-    ] @ Option.cata (fun body -> ["body", `String (Pp.string_of_ppcmds @@ pr_lconstr_env ~inctx:true env sigma body)]) [] body) in
-    (env', hyp :: l)
+  let rec loop (events : 'a LspManager.events) =
+    let open Yojson.Basic.Util in
+    log @@ "[T] looking for next step";
+    Lwt_io.flush_all () >>= fun () ->
```
Comment (gares) [inline ide/vscoq/vscoqtop.ml:21]: This is the final "main loop" which does not look too ugly to me

File: ide/vscoq/vscoqtop_worker.ml [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  [ "-vscoqtop_master"; port ] -> int_of_string port, []
+  | _ -> assert false
+
+let islave_default_opts =
+  Coqargs.default
+
+let islave_init run_mode ~opts =
+  Coqtop.init_toploop opts
+
+let main () =
```
Comment (gares) [inline ide/vscoq/vscoqtop_worker.ml:40]: This toplevel is needed when we can't fork. We could have just one toplevel and use command flags to choose which role it plays, but we would not save much.

File: stm/delegationManager.ml [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+let log msg = Format.eprintf "%d] @[%s@]@\n%!" (Unix.getpid ()) msg
+
+type sentence_id = Stateid.t
+
+type link = {
+  write_to :  Lwt_io.output_channel;
+  read_from:  Lwt_io.input_channel;
+}
+
+(* TODO: move promises in the non-marshalable part to be shure *)
```
Comment (gares) [inline stm/delegationManager.ml]:
```suggestion
```
done

File: stm/delegationManager.mli [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+(* Event for the main loop *)
+type event
+type 'a events = ([> `DelegationManager of event ] as 'a) Lwt.t list
+
+val handle_event : event -> 'a events Lwt.t
+
+(* When a worker is available [job] is called and when it returns the
+   event becomes ready; in turn the event triggers the action *)
+val worker_available :
```
Comment (gares) [inline stm/delegationManager.mli]: This is the *one* API to delegate.

**Comment by gares** [comment]
There seem to be a bug in GH since I don't see the most recent commits I've on my branch here...

**Comment by gares** [comment]
I had a minute to better separate the concerns of the delegation manager and the execution manager. It is not 100% type safe (it returns a `'job`), but the plumbing is completely hidden. ExecManager has now 2 init, one of which is not only creates a state but also a remote mapping for a given list of sentences that were in the original remote mapping.

File: ide/vscoq/vscoqtop.ml (hunk 1 above)
Comment (gares) [reply ide/vscoq/vscoqtop.ml:21]: The alternative is to hide sub components' events inside their super component.
In this way we don't need polymorphic variants, but you would call lspmanager.handle_event always, and in some cases that handle event would dispatch to other handle_event (like the one of the delegation manager or the one of the tactic manager if we implement par:).

Comment (gares) [reply ide/vscoq/vscoqtop.ml:21]: I've implemented the sealing, the main loop is now agnostic on the component/type of events

**Comment by gares** [comment]
I've added 2 infix:
- `>?=` that returns a result, so that you can see if the promise was fulfilled with an error an act accordingly. Only used by the worker manager, that prints an error and terminates when a worker ends.
- `>!=` asserts no error took place. It is used whenever some i/o is performed, but the type does not enforce this :-/ In particular it print unix errors and the like and then terminates the process. If you forge it the exception can raise the toplevel.

I'm not super happy about that, I need to think more. Lwt puts the exception monad outside, so ideally you don't see it. It would nicer to have all i/o related bind automagically print the message and the file location. Weather to exit 1 or re-raise the exception is to be understood. To me 99% of the times you should abort. The only exception seems to be the worker manager call to read.

**Comment by maximedenes** [comment]
I've been trying to rebase this branch for half a day now. Do the commits build?

**Comment by maximedenes** [comment]
I must be doing something wrong, but I even fail to build one of the first commits (0af5ca4e00989f3690c36d8af5fad2f875bff637). I probably lost a lot of time trying to understand how I had screwed up the rebase. What am I supposed to do to build it correctly?

**Comment by gares** [comment]
Up to "cancel worker" they should

**Comment by gares** [comment]
I'll try to revive this tomorrow

**Comment by maximedenes** [comment]
@gares I still fail to build 659cce2a7b3e287966161797d0abf875f08c31e5

Are you sure you pushed the version that you fixed the other day?

**Comment by gares** [comment]
I probably forgot, I'll do it tomorrow

**Pull request closed without merging**